r/R/dataset.R

```diff
@@ -43,19 +43,36 @@
 #'   * `NULL` for no partitioning
 #'
 #' The default is to autodetect Hive-style partitions.
+#' @param unify_schemas logical: should all data fragments (files, `Dataset`s)
+#' be scanned in order to create a unified schema from them? If `FALSE`, only
+#' the first fragment will be inspected for its schema. Use this
+#' fast path when you know and trust that all fragments have an identical schema.
+#' The default is `FALSE` when creating a dataset from a file path (because
+#' there may be many files and scanning may be slow) but `TRUE` when `sources`
+#' is a list of `Dataset`s (because there should be few `Dataset`s in the list
+#' and their `Schema`s are already in memory).
 #' @param ... additional arguments passed to `dataset_factory()` when
 #' `sources` is a file path, otherwise ignored.
 #' @return A [Dataset] R6 object. Use `dplyr` methods on it to query the data,
 #' or call [`$NewScan()`][Scanner] to construct a query directly.
 #' @export
 #' @seealso `vignette("dataset", package = "arrow")`
 #' @include arrow-package.R
-open_dataset <- function(sources, schema = NULL, partitioning = hive_partition(), ...) {
+open_dataset <- function(sources,
+                         schema = NULL,
+                         partitioning = hive_partition(),
+                         unify_schemas = NULL,
+                         ...) {
   if (is_list_of(sources, "Dataset")) {
     if (is.null(schema)) {
-      # Take the first one.
-      # Someday, we should expose a way to unify schemas
-      schema <- sources[[1]]$schema
+      if (is.null(unify_schemas) || isTRUE(unify_schemas)) {
+        # Default is to unify schemas here
+        schema <- unify_schemas(schemas = map(sources, ~.$schema))
```

Comment:
I guess this is probably a lack of R knowledge on my part, but it looks extremely odd to have declared `unify_schemas` as a parameter of this function then later use an identically named function. I would have expected the function `unify_schemas` to be shadowed by the function parameter

Reply:
Yeah I think R lookup works different when you’re calling a function